### DISCUSSION TITLE
Project panel from lexicographical to natural sorting.

### DISCUSSION THREAD

**Discussion by TomPlanche** [zed-industries/zed#20126] [Show and tell]
While working on Leetcode problems, I stumbled upon a project panel sorting who bothered me.
The folders (and files) are lexicographicaly sorted...

<img width="130" alt="Screenshot 2024-11-03 at 06 07 32" src="https://github.com/user-attachments/assets/7a63c4d6-5999-48ef-8163-026bafb61865">
<img width="115" alt="Screenshot 2024-11-03 at 06 11 54" src="https://github.com/user-attachments/assets/a81aa148-b122-4d3f-aae3-e6a63d16e3e4">

This is annoying and not natural.

I found that the responsible for this is `sort_worktree_entries` in the [`crates/project/src/project.rs`](https://github.com/zed-industries/zed/tree/main/crates/project_panel/src/project_panel.rs) file used in the line 2005 `project::sort_worktree_entries(&mut visible_worktree_entries);` of the [`./project_panel/src/project_panel.rs`](https://github.com/zed-industries/zed/tree/main/crates/project_panel/src/project_panel.rs) file.
This uses the [`./crates/util/src/paths.rs`](https://github.com/zed-industries/zed/tree/main/crates/util/src/paths.rs)'s `compare_paths` function.

I'm changing the `compare_paths` function to `compare_paths_with_sort_mode`.

Now the folders and files are in alphabetical order.

<img width="185" alt="Screenshot 2024-11-03 at 06 20 39" src="https://github.com/user-attachments/assets/48f70d08-d12a-4858-b7eb-2bfc93e3277f">
<img width="211" alt="Screenshot 2024-11-03 at 06 20 49" src="https://github.com/user-attachments/assets/bfd7edbf-c5be-4a98-9128-f76ecd87402f">

**Comment by SomeoneToIgnore** [comment]
You're definitely not supposed to change things forever, but simply replacing the sorting mechanism: someone prefers things the other way around.

What's really needed is multiple sorting modes and a way to toggle between them.

**Reply by TomPlanche** [reply]
I've made a POC [here](https://github.com/TomPlanche/zed/tree/better-project-panel).
How can I improve this ?
I created the `crates/sort_strategies` because I couldn't manage to make the `SortStrategy` enum public and import it where I needed it without dependency circles.

**Reply by SomeoneToIgnore** [reply]
Nice, thank you.
My notes, no particular order:

* `Cargo.lock` shows a few dependencies bumped, those changes seem unrelated and are not welcome to be mixed into all this

* The `test_natural_sort_order` is good but:
1. it's not in the right place, let's make it more unit-testy and use less dependencies for it, moving the test near the method that sorts paths. I have thought we have a test in `util` already, nearby the `compare_paths`? 
2. each new sorting approach is better be tested

* Creating another crate for an enum seems like an overkill. Not quite sure I understand what kind of cycle there will be if we put it into `utils`, in the `paths` module? `util` crate should be a leaf one, it does not depend on anything that uses it

* Both *.md and *.json config examples need to enumerate all possibilities

* `pub fn sort_worktree_entries_with_strategy` is very odd, as the only usage seems to be in the project panel?
I would inline it and do not expose another public API point in `project`.

* This new setting has to be applied uniformly across the codebase:
1. `outline_panel` needs the same 
2. project search needs the same
3. whatever else deals with multi buffer entry sorting needs the same: it would be very confusing to get one sorting in the panel, and another in the multi buffer

Basically everything that uses `compare_paths` has to stop doing it and use the same setting and the new method instead.
Also, it's crucial to test it here, otherwise Rust's stdlib may panic: https://github.com/zed-industries/zed/pull/17604

* `Add more strategies here as needed` and other stray comments are useless, obvious and not needed

**Reply by TomPlanche** [reply]
@SomeoneToIgnore Hi, I've been away from this project, sorry.

I'm redoing the work and I'm almost finished: [branch here](https://github.com/TomPlanche/zed/tree/better-sorting).

I'm just having trouble to get rid of the `/crate/project`'s `compare_paths` because of cycle dependency, WIP.

**Reply by SomeoneToIgnore** [reply]
No worries, it's open source and no one expects good timings from contributors.

I still do not understand what the issue is, as 

> `/crate/project`'s `compare_paths` 


seems wrong to me, the function is in `crates/util` which is a leaf crate and does not depend on anything, everybody uses it instead.

**Reply by TomPlanche** [reply]
I have a:
```rs
#[derive(Copy, Clone, Debug, Default, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SortMode {
    #[default]
    Lexicographical,
    Alphabetical,
}
```
in the `crates/util/src/paths.rs`, this cause no problem.

I use it in the `compare_paths_with_sort_mode` function in the `crates/util/src/paths.rs`.
This function will replace the old `compare_paths`.

But, in order to use this enum, I place it in the `crates/project_panel/src/project_panel_settings.rs` to use it in the `settings.json` file:
```rs
#[derive(Deserialize, Debug, Clone, Copy, PartialEq)]
pub struct ProjectPanelSettings {
    // ... 
    pub show_diagnostics: ShowDiagnostics,
    pub sort_mode: SortMode, // here
}
```

I use this `ProjectPanelSettings` where I need to fetch the settings from the `cx` like so:
```rs
let sort_mode = ProjectPanelSettings::get_global(cx).sort_mode;
```

So I need to have the `ProjectPanelSettings` available so the `crate/project_panel` dependency.

All `compare_path` have been replaced except one: `crates/project/src/project.rs`'s since `crates/project_panel` includes `crates/project` in its dependencies, thus creating a cycle dependency.

**Reply by SomeoneToIgnore** [reply]
So, looks like `ProjectPanelSettings` is the wrong place to put this into?
Sounds odd indeed, as we're supposed to sort things in the outline panel at least, same way as in the project panel?
Seems that you want it in the file_finder's `open_path_prompt.rs` which is, again, not the `ProjectPanel`.

* `ProjectSettings` could be a better place?
* We might reconsider "sort everything similarly" restriction from my previous write-ups, as things have mutated a bit since the last time: `open_path_prompt.rs` might want a different sorting? Not sure but worth checking, see https://github.com/zed-industries/zed/pull/25045  for more details.

**Reply by TomPlanche** [reply]
`ProjectSettings` looks good, made the changes, does it look good to you ?

**Reply by TomPlanche** [reply]
We have a dedicated `file_sorting` key that for now only takes a `strategy` field.
We could add new strategies, dedicated strategies per panel, etc.

I thought about adding a `file_before_folders` boolean, a `hidden_file_first` boolean, `uppercase_before_lowercase` for greater personalization.

**Reply by SomeoneToIgnore** [reply]
That can also go as a separate setting to each panel, in theory, but this all are details, seems that you're over your main issues and ready to submit a PR?

**Reply by TomPlanche** [reply]
PR live at #25148 :)

**Comment by jnptk** [comment]
What's the status here? Really looking forward for this feature!

**Reply by TomPlanche** [reply]
The choice varies between making alphabetical sorting the default value or making it a parameter.
#25148.
